src/new/sendTransaction.ts

```diff
@@ -494,4 +509,162 @@ class SendTransaction extends EventEmitter {
+    newtxData.inputs = newUtxos.utxos.map((utxo) => {
+      return {
+        txId: utxo.txId,
+        index: utxo.index,
+        value: utxo.value,
+        authorities: utxo.authorities,
+        token: utxo.token,
+        address: utxo.address,
+      } as IDataInput;
+    });
+
+    if (newUtxos.amount > outputAmount) {
+      // We need to create a change output
+      const changeAddress = options.changeAddress || await storage.getCurrentAddress();
+      const changeOutput: IDataOutput = {
+        type: 'p2pkh', //get from wallet
+        token,
+        value: newUtxos.amount - outputAmount,
+        address: changeAddress,
+        authorities: 0,
+        timelock: null,
+        isChange: true,
+      };
+      newtxData.outputs.push(changeOutput);
+    }
+  } else {
+    let inputAmount = 0;
+    for (const input of dataTx.inputs) {
+      // We will check the validity and availability of the provided inputs
+      // and the amount (suggesting a change if needed)
+      // The inputs do not need to be added on newtxData.inputs since they are provided by the caller.
+      const checkSpent = await checkUnspentInput(storage, input, token);
+      if (!checkSpent.success) {
+        throw new Error(`Token: ${token}. ${checkSpent.message}`);
+      }
+
+      if (!transactionUtils.canUseUtxo(input, storage)) {
+        throw new Error(`Token: ${token}. Output [${input.txId}, ${input.index}] is locked or being used`);
+      }
+
+      inputAmount += input.value;
+    }
+    if (inputAmount < outputAmount) {
+      throw new Error(`Token: ${token}. Sum of outputs is greater than sum of inputs`);
+    }
+    if (inputAmount > outputAmount) {
+      // Need to create a change output
+        const changeAddress = options.changeAddress || await storage.getCurrentAddress();
+        newtxData.outputs.push({
+        type: 'p2pkh', //get from wallet
+        token,
+        value: inputAmount - outputAmount,
+        address: changeAddress,
+        authorities: 0,
+        timelock: null,
+        isChange: true,
+      });
+    }
+  }
+
+  return newtxData;
+}
+
+/**
+ * Check that the input is unspent, valid and available.
+ * Will return a user-friendly message if it is not.
+ *
+ * @param {IStorage} storage The storage instance
+ * @param {IDataInput} input The input we are checking
+ * @param {string} selectedToken The token uid we are checking
+ * @returns {Promise<{success: boolean, message: string}>}
+ */
+export async function checkUnspentInput(
+  storage: IStorage,
+  input: IDataInput,
+  selectedToken: string,
+): Promise<{success: boolean, message: string}> {
+  const tx = await storage.getTx(input.txId);
+  if (tx === null) {
+    return { success: false, message: `Transaction [${input.txId}] does not exist in the wallet` };
+  }
+  if (tx.is_voided) {
+    return { success: false, message: `Transaction [${input.txId}] is voided` };
+  }
+  if (tx.outputs.length - 1 > input.index) {
```

Comment:
Shouldn't it be the opposite? 

```suggestion
  if (tx.outputs.length - 1 < input.index) {
```